hasher-matcher-actioner/hmalib/common/timebucketizer.py

```diff
@@ -174,8 +178,53 @@ def get_records(
 
         for file in file_list:
             with open(file, "r") as my_file:
-                content_list.extend(
-                    list(map(type_class().from_csv, csv.reader(my_file)))
-                )
+                content_list.extend(list(map(type_class.from_csv, csv.reader(my_file))))
 
         return content_list
+
+    @staticmethod
+    def squash_content(
+        location: datetime.datetime,
+        type: str,
+        storage_path: str,
+        bucket_width: datetime.timedelta,
+        date_to: datetime.timedelta = datetime.timedelta(days=1),
+    ):
+
+        if bucket_width > date_to:
+            return
+
+        until_nearest = TimeBucketizer._calculate_bucket_endpoints(
+            location, bucket_width
+        )[1]
+
+        since_nearest = TimeBucketizer._calculate_bucket_endpoints(
+            until_nearest - date_to, bucket_width
+        )[0]
+
+        while until_nearest > since_nearest:
+
+            directory_path = TimeBucketizer._generate_path(
+                storage_path, type, until_nearest
+            )
+
+            if os.path.isdir(directory_path):
+                file_list = []
+                for file in os.listdir(directory_path):
+                    file_path = os.path.join(directory_path, file)
+                    if file.startswith("squash"):
+                        return
+                    elif os.path.isfile(file_path):
+                        file_list.append(file_path)
+
+                with open(
+                    os.path.join(directory_path, "squash" + str(uuid.uuid1())) + ".csv",
+                    "w",
+                ) as new_file:
+                    for file in file_list:
+                        with open(file, "r") as reader:
+                            new_file.write(reader.read())
+
+                        os.remove(file)
+
+            until_nearest -= bucket_width
```

Comment:
I would suggest you break down the problem into two parts: a) Identify the buckets to squash b) actually squash.

A problem with the current approach is if the loop gets terminated prematurely because of an error or a timeout. It is impossible to know what buckets could not be squashed. 

eg. time increases as we go down: 
1. squashed_in_an_earlier_run
2. squashed_in_an_earlier_run 
3. squashed_in_an_earlier_run,
4. errored 
5. errored
6. squashed_in_current_run 
7. squashed_in_current_run
8. too_new_to_squash.

Instead, run your 'identify loop', build a list of all buckets to process, and then process them chronologically. So, even in case of an error, you end up with.

1. squashed_in_an_earlier_run
2. squashed_in_an_earlier_run 
3. squashed_in_an_earlier_run,
4. squashed_in_current_run 
5. squashed_in_current_run
6. errored 
7. errored
8. too_new_to_squash.